### ISSUE TITLE
Bugs for run_electra.py at master branch

### ISSUE THREAD

**Issue by ZiyueHuang** [dmlc/gluon-nlp#1315]
### Bug for Horovod support

At https://github.com/dmlc/gluon-nlp/blob/master/scripts/pretraining/run_electra.py#L418, `num_samples_per_update` depends on the data batch, and thus could be different across the horovod processes (note that Horovod will launch a process for each GPU card), meaning that the parameters maintained in each horovod process (since we use `trainer.allreduce_grads()`) might diverge.

We can normalize the loss by batch_size on each worker before `trainer.allreduce_grads()`, to avoid syncing `num_samples_per_update` across all workers.

### Bug for MLM loss

In the current implementation, the MLM loss actually doesn't take into account `masked_weights`.

At https://github.com/dmlc/gluon-nlp/blob/master/scripts/pretraining/run_electra.py#L383-L384, `masked_weights.reshape(-1)` should be replaced by `mx.np.expand_dims(masked_weights, axis=-1)`, please see the doc of `SoftmaxCELoss` at https://github.com/apache/incubator-mxnet/pull/19010/files or the example usage in bert at https://github.com/dmlc/gluon-nlp/blob/v0.10.x/scripts/bert/pretraining_utils.py#L416-L418

Note that

`mlm_scores`'s shape: (batch_size, num_masked_positions, vocab_size)
`unmasked_tokens`'s shape: (batch_size, num_masked_positions)
`masked_weights`'s shape: (batch_size, num_masked_positions)

If we comment out `mlm_loss_fn.hybridize()` and add two lines before/after https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/gluon/loss.py#L408: `print('loss.shape before weighting ', loss.shape, ' sample_weight.shape ', sample_weight.shape)` and `print('loss.shape after weighting ', loss.shape)`, running the current script run_electra.py we will get

```
loss.shape before weighting  (64, 19, 1)  sample_weight.shape  (1216,)
loss.shape after weighting  (64, 19, 1216)
```
note that 1216 = 64 x 19

cc @sxjscience @ZheyuYe 

### Error Message
(Paste the complete error message, including stack trace.)

## To Reproduce
(If you developed your own code, please provide a short script that reproduces the error. For existing examples, please provide link.)

### Steps to reproduce
(Paste the commands you ran that produced the error.)

1.
2.

## What have you tried to solve it?

1.
2.

## Environment

We recommend using our script for collecting the diagnositc information. Run the following command and paste the outputs below:
```
curl --retry 10 -s https://raw.githubusercontent.com/dmlc/gluon-nlp/master/tools/diagnose.py | python

# paste outputs here
```

**Comment by ZheyuYe** [comment]
There is indeed a Bug for mlm loss here, thank you very patiently and carefully to revise the training script and find this bug out. As Far as I can see, this MLM bug will lead to a steep increase in calculations which need to be fixed.

**Comment by ZiyueHuang** [comment]
@ZheyuYe Thanks for the response, I will work on the fix later.

**Comment by sxjscience** [comment]
That might also explain why it's slower. Thanks for the observation! We need to refactor the pretraining script to make it more structured.
